Make `RustString` an extern type to avoid `improper_ctypes` warnings

Zalathar opened the pull request:
Currently, any FFI function that uses `&RustString` needs to also add `#[ignore(improper_ctypes)]` to silence a warning.

The warning is not _completely_ bogus, because `RustString` contains `Vec<u8>` and therefore does not have a guaranteed layout. But we have no way of telling the lint that this doesn't matter, because the C++ code only uses that pointer opaquely and never relies on its underlying layout.

Ideally there would be some way to silence `improper_ctypes` at the type-definition site. But because there isn't, casting to and from a separate extern type is better than having to annotate every single use site.

Zalathar committed:
Make `RustString` an extern type to avoid `improper_ctypes` warnings

Zalathar commented:
I also want to rename `RustString` to something else, because it's a raw byte buffer rather than a UTF-8 sequence, but that would be too much scope creep for this PR.

Zalathar commented:
See also #116963 for general dissatisfaction with `improper_ctypes`.

Zalathar commented:
I'm not 100% confident that casting between normal references and extern type references like this is OK, but it seems reasonable?

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/11647904759/job/32433704881) [(plain)](https://github.com/rust-lang/rust/commit/5e35d13b2eaf822d296583afaf2e1b0d9dd0b8fa/checks/32433704881/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#16 3.171 Building wheels for collected packages: reuse
#16 3.172   Building wheel for reuse (pyproject.toml): started
#16 3.419   Building wheel for reuse (pyproject.toml): finished with status 'done'
#16 3.420   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132720 sha256=026f3bb0f1aa8090b861fd0a0939cb1a782396d84c8aab7875096557d637a0f6
#16 3.420   Stored in directory: /tmp/pip-ephem-wheel-cache-c362_14d/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#16 3.423 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#16 3.826 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#16 3.827 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#16 DONE 3.9s
---
 Documenting rustc_hir v0.0.0 (/checkout/compiler/rustc_hir)
error: unresolved link to `&RustString`
  --> compiler/rustc_llvm/src/lib.rs:18:67
   |
18 |     /// in conjunction with `RawRustStringOstream`. Use this as [`&RustString`]
   |                                                                   ^^^^^^^^^^^ no item named `&RustString` in scope
   = help: to escape `[` and `]` characters, add '\' before them like `\[` or `\]`
   = note: `-D rustdoc::broken-intra-doc-links` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(rustdoc::broken_intra_doc_links)]`


error: unresolved link to `&RustString`
  --> compiler/rustc_llvm/src/lib.rs:67:49
   |
67 | /// Appends the contents of a byte slice to a [`&RustString`].
   |                                                 ^^^^^^^^^^^ no item named `&RustString` in scope
   = help: to escape `[` and `]` characters, add '\' before them like `\[` or `\]`

error: could not document `rustc_llvm`
warning: build failed, waiting for other jobs to finish...
warning: build failed, waiting for other jobs to finish...
Command has failed. Rerun with -v to see more details.
  local time: Sun Nov  3 03:21:23 UTC 2024
  network time: Sun, 03 Nov 2024 03:21:24 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

Zalathar commented:
See https://github.com/rust-lang/rust/pull/67077/commits/47e932b96e726d855f9885f073ef6b6b6bb78438 for why RustString is in `rustc_llvm`.

Zalathar committed:
Don't use `LLVMRustStringWriteImpl` outside of `RawRustStringOstream`

Zalathar commented:
Rebased to remove another `ignore` added by #132514 [(diff)](https://github.com/rust-lang/rust/compare/a1dcb0221669561060105950e9c8a821c9390842..730d08819d958b806f0117da427d5805202fdfc6).

Zalathar commented:
Rebased to resolve conflicts and remove more ignore/expect added by (my) other PRs.

Zalathar commented on compiler/rustc_llvm/src/lib.rs:
```diff
@@ -2,42 +2,81 @@
 #![warn(unreachable_pub)]
 // tidy-alphabetical-end
 
-// NOTE: This crate only exists to allow linking on mingw targets.
```
I'm not sure what this was originally referring to, but it's pretty clearly no longer true of the present crate.

cuviper reviewed (changes requested):
Overall it looks like a nice improvement! I just have a couple quibbles.

Zalathar commented:
Addressed review feedback [(diff)](https://github.com/rust-lang/rust/compare/b266f0c9519b464247a2635f3ed3c06bef255dc7..89d7efaf8f5ed8c1c15faea21824ae479656bdc1).

Zalathar commented on compiler/rustc_llvm/src/lib.rs:
```diff
@@ -2,42 +2,81 @@
+        // `RustStringInner`.
+        let ptr: *const RustString = ptr::from_ref(opaque);
+        let ptr: *const RustStringInner = ptr.cast();
+        unsafe { ptr.as_ref() }.unwrap()
```
Yeah, I must have been influenced by other code where the null case is relevant, but here we start with a reference so the unwrap doesn't add anything.

Zalathar commented:
@rustbot ready

cuviper commented:
@bors r+

bors commented:
:pushpin: Commit 89d7efaf8f5ed8c1c15faea21824ae479656bdc1 has been approved by `cuviper`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=cuviper 89d7efaf8f5ed8c1c15faea21824ae479656bdc1 -->
<!-- homu: {"type":"Approved","sha":"89d7efaf8f5ed8c1c15faea21824ae479656bdc1","approver":"cuviper","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 89d7efaf8f5ed8c1c15faea21824ae479656bdc1 with merge 012ae13d6ac41fb4df649d39979f39b08241a290...
<!-- homu: {"type":"BuildStarted","head_sha":"89d7efaf8f5ed8c1c15faea21824ae479656bdc1","merge_sha":"012ae13d6ac41fb4df649d39979f39b08241a290"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/11753407958/job/32748103051)
Approved by: cuviper
Pushing 012ae13d6ac41fb4df649d39979f39b08241a290 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"cuviper","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/11753407958/job/32748103051"},"merge_sha":"012ae13d6ac41fb4df649d39979f39b08241a290"} -->

rust-timer commented:
Finished benchmarking commit (012ae13d6ac41fb4df649d39979f39b08241a290): [comparison URL](https://perf.rust-lang.org/compare.html?start=328b759142ddeae96da83176f103200009d3e3f1&end=012ae13d6ac41fb4df649d39979f39b08241a290&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=328b759142ddeae96da83176f103200009d3e3f1&end=012ae13d6ac41fb4df649d39979f39b08241a290&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=328b759142ddeae96da83176f103200009d3e3f1&end=012ae13d6ac41fb4df649d39979f39b08241a290&stat=max-rss)
<details>
<summary>Results (secondary 3.1%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 3.1% | [2.5%, 4.3%] | 4     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=328b759142ddeae96da83176f103200009d3e3f1&end=012ae13d6ac41fb4df649d39979f39b08241a290&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

### [Binary size](https://perf.rust-lang.org/compare.html?start=328b759142ddeae96da83176f103200009d3e3f1&end=012ae13d6ac41fb4df649d39979f39b08241a290&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 780.614s -> 780.969s (0.05%)
**Artifact size**: 335.33 MiB -> 335.32 MiB (-0.00%)
